modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +51,146 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+        req
+          .as[Signed[DataUpdate]]
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+            }
+          }.flatMap { results =>
+            val errors = results.collect { case Left(_) => InvalidSignature.toApplicationError }
+            val validTransactions = results.collect { case Right(hashed) => hashed }
+
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
+              )
+            } else if (validTransactions.isEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("No valid data transactions found"))
+              )
+            } else {
+              NonEmptyList.fromList(validTransactions.map(_.hash)) match {
+                case Some(hashes) =>
+                  dataTransactionsQueue
+                    .offer(NonEmptyList.fromListUnsafe(validTransactions.map(_.signed)))
+                    .flatMap { _ =>
+                      val (dataUpdateHashes, feeTransactionHashes) = validTransactions
+                        .partition(_.signed.value.isInstanceOf[DataUpdate])
+
+                      if (feeTransactionHashes.isEmpty) {
+                        Ok(("hash" ->> dataUpdateHashes.map(_.hash.value)) :: HNil)
+                      } else {
+                        Ok(
+                          ("hashes" ->> dataUpdateHashes.map(_.hash.value)) ::
+                            ("feeHash" ->> feeTransactionHashes.map(_.hash.value)) ::
+                            HNil
+                        )
+                      }
+                    }
+                case None =>
+                  BadRequest(
+                    Json.obj("error" -> Json.fromString("No valid data transactions found"))
+                  )
               }
+            }
           }
         }
-        .handleUnknownError
+
+      def handleDataUpdate(dataUpdate: Signed[DataUpdate]): F[Response[F]] = {
+        val dataTransactions: DataTransactions = NonEmptyList.one(dataUpdate.widen[DataTransaction])
```

Comment:
you don't need an explicit type if you use widen.

Reply:
👍🏼